sound/soc/sof/sof-client-fw-gdb.c

```diff
@@ -0,0 +1,333 @@
+#include <linux/debugfs.h>
+#include <linux/ktime.h>
+#include <linux/mod_devicetable.h>
+#include <linux/module.h>
+#include <linux/pm_runtime.h>
+#include <linux/slab.h>
+#include <linux/uaccess.h>
+#include <sound/sof/header.h>
+#include <sound/sof/ipc4/header.h>
+
+#include "sof-client.h"
+#include "sof-priv.h"
+#include "ops.h"
+
+#define SOF_IPC_CLIENT_SUSPEND_DELAY_MS	3000
+
+struct sof_fw_gdb_priv {
+	struct dentry *dfs_file;
+	size_t max_msg_size;
+	enum sof_ipc_type ipc_type;
+
+	struct sof_ipc_gdb_message *tx_buffer;
+	void *rx_buffer;
+};
+
+struct sof_ipc_gdb_message {
+	struct sof_ipc_cmd_hdr hdr;
+	char cmd[];
+} __packed;
+
+#define RING_SIZE 128
+struct ring {
+	unsigned char head;
+	unsigned char fill1[63];
+	unsigned char tail;
+	unsigned char fill2[63];
+	unsigned char data[RING_SIZE];
+} __packed;
+
+static int sof_fw_gdb_send_message(struct sof_client_dev *cdev)
+{
+	struct sof_fw_gdb_priv *priv = cdev->data;
+	struct device *dev = &cdev->auxdev.dev;
+	int ret, err;
+
+	ret = pm_runtime_resume_and_get(dev);
+	if (ret < 0 && ret != -EACCES) {
+		dev_err_ratelimited(dev, "debugfs write failed to resume %d\n", ret);
+		return ret;
+	}
+
+	/* send the message */
+	ret = sof_client_ipc_tx_message(cdev, priv->tx_buffer, priv->rx_buffer,
+					priv->max_msg_size);
+	if (ret)
+		dev_err(dev, "IPC message send failed: %d\n", ret);
+
+	pm_runtime_mark_last_busy(dev);
+	err = pm_runtime_put_autosuspend(dev);
+	if (err < 0)
+		dev_err_ratelimited(dev, "debugfs write failed to idle %d\n", err);
+
+	return ret;
+}
+
+static bool gdb_has_init = false;
+
+static int sof_fw_gdb_dfs_open(struct inode *inode, struct file *file)
+{
+	struct sof_client_dev *cdev = inode->i_private;
+	struct sof_fw_gdb_priv *priv = cdev->data;
+	int ret;
+
+	if (sof_client_get_fw_state(cdev) == SOF_FW_CRASHED)
+		return -ENODEV;
+
+	ret = debugfs_file_get(file->f_path.dentry);
+	if (unlikely(ret))
+		return ret;
+
+	ret = simple_open(inode, file);
+	if (ret)
+		debugfs_file_put(file->f_path.dentry);
+
+	// When file is opened, send GDB init command to FW
+	priv->tx_buffer->hdr.cmd = SOF_IPC_GLB_GDB_DEBUG;
+	priv->tx_buffer->hdr.size = sizeof(struct sof_ipc_cmd_hdr);
+	if (!gdb_has_init) {
+		sof_fw_gdb_send_message(cdev);
+		gdb_has_init = true;
+	}
+
+	return ret;
+}
+
+static int sof_debug_read(struct snd_sof_dev *sdev,
+			  char *buff, size_t bytes)
+{
+	int count = 0;
+	int bar = snd_sof_dsp_get_bar_index(sdev, SOF_FW_BLK_TYPE_SRAM);
+	void __iomem *rx_ring_buffer = sdev->bar[bar] + sdev->debug_box.offset + 256;
+	char* msg_region = rx_ring_buffer + 128;
+
+
+	uint8_t head = readb(rx_ring_buffer);
+	uint8_t tail = readb(rx_ring_buffer + 64);
+
+	for (int i = 0; i < bytes; i++) {
+		if (tail == head) {
+			// No Data
+			break;
+		}
+		buff[i] = readb(msg_region + tail);
+		tail = (tail + 1) % RING_SIZE;
+		count++;
+	}
+
+	writeb(tail, rx_ring_buffer + 64);
+
+	return count;
+}
+
+static ssize_t sof_fw_gdb_dfs_read(struct file *file, char __user *buffer,
```

Comment:
we need a lot of comments explaining how we tunnel GDB commands to IPC4 and vice versa.